pkg/fakedri/fakedri.go

```diff
@@ -0,0 +1,529 @@
+	topology := opts.Capabilities["connection-topology"]
+	gpus := opts.DevCount
+	tiles := opts.TilesPerDev
+	connections := opts.Capabilities["connections"]
+
+	if topology == fullyConnected {
+		saveSideCarFile(buildConnectionList(gpus, tiles))
+	} else if connections != "" {
+		saveSideCarFile(connections)
+	} else {
+		return
+	}
+
+	klog.V(1).Infof("XELINK: generated xelink sidecar label file, using (GPUs: %d, Tiles: %d, Topology: %s)", gpus, tiles, topology)
+}
+
+func buildConnectionList(gpus, tiles int) string {
+	var nodes = make([]string, 0)
+
+	for mm := 0; mm < gpus; mm++ {
+		for nn := 0; nn < tiles; nn++ {
+			nodes = append(nodes, fmt.Sprintf("%d.%d", mm, nn))
+		}
+	}
+
+	var links = make(map[string]bool, 0)
+
+	var smap = make([]string, 0)
+
+	for _, from := range nodes {
+		for _, to := range nodes {
+			if to == from {
+				continue
+			}
+
+			link := fmt.Sprintf("%s-%s", to, from)
+
+			reverselink := fmt.Sprintf("%s-%s", from, to)
+			if _, exists := links[reverselink]; !exists {
+				links[link] = true
+
+				smap = append(smap, link)
+			}
+		}
+	}
+
+	return strings.Join(smap, "_")
+}
+
+func saveSideCarFile(connections string) {
+	// Get user-specific temp directory
+	filePath := filepath.Join("/etc/kubernetes/node-feature-discovery/features.d", "xpum-sidecar-labels.txt")
+
+	// Safely create file in the temp directory
+	f, err := os.Create(filePath)
+	if err != nil {
+		klog.Fatalf("Failed to create file: %v", err)
+	}
+	defer f.Close()
+
+	line := fmt.Sprintf("xpumanager.intel.com/xe-links=%s", connections[:min(len(connections), maxK8sLabelSize)])
+	klog.V(1).Info(line)
+
+	if _, err := f.WriteString(line + "\n"); err != nil {
+		panic(err)
+	}
+
+	index := 2
+
+	for i := maxK8sLabelSize; i < len(connections); i += (maxK8sLabelSize - 1) {
+		line := fmt.Sprintf("xpumanager.intel.com/xe-links%d=Z%s", index, connections[i:min(len(connections), i+maxK8sLabelSize-1)])
+		klog.V(1).Info(line)
+
+		if _, err := f.WriteString(line + "\n"); err != nil {
+			panic(err)
+		}
+
+		index++
+	}
+}
+
+func MakeOptions(opts GenOptions) GenOptions {
+	if opts.DevCount < 1 || opts.DevCount > maxDevs {
+		klog.Fatalf("Invalid device count: 1 <= %d <= %d", opts.DevCount, maxDevs)
+	}
+
+	if opts.VfsPerPf > 0 {
+		if opts.TilesPerDev > 0 || opts.DevsPerNode > 0 {
+			klog.Fatalf("SR-IOV VFs (%d) with device tiles (%d) or Numa nodes (%d) is unsupported for faking",
+				opts.VfsPerPf, opts.TilesPerDev, opts.DevsPerNode)
+		}
+
+		if opts.DevCount%(opts.VfsPerPf+1) != 0 {
+			klog.Fatalf("%d devices cannot be evenly split to between set of 1 SR-IOV PF + %d VFs",
+				opts.DevCount, opts.VfsPerPf)
+		}
+	}
+
+	if opts.DevsPerNode > opts.DevCount {
+		klog.Fatalf("DevsPerNode (%d) > DevCount (%d)", opts.DevsPerNode, opts.DevCount)
+	}
+
+	if opts.DevMemSize%mib != 0 {
+		klog.Fatalf("Invalid memory size (%f mib), not even mib", float64(opts.DevMemSize)/mib)
+	}
+
+	return opts
+}
+
+func GetOptions(name string) GenOptions {
```

Comment:
Redundantly moving function to a different place in file makes reviewing harder...

Reply:
replaced with 2 functions:
GetOptionsByJSON
GetOptionsByYAML